### PR TITLE
[Optimize](functions)Optimize function call for const columns.

### PR DESCRIPTION
**Pull request by zclllyybb**
# Proposed changes

Since there are many function tasks under construction, it's necessary to standardize the behavior of convert a column may be const to a concrete column. 
The basic method to replace previous usgae `convert_to_full_column_if_const` is like:
```c++
ColumnPtr argument_columns[2];
bool col_const[2];
for (int i = 0; i < 2; ++i) {
    std::tie(argument_columns[i], col_const[i]) =
            unpack_if_const(block.get_by_position(arguments[i]).column);
    check_set_nullable(argument_columns[i], null_map); // set null map if you need.
}
```
Then, specialize the function like `vector_scalar` to distinguish the implementation by `col_const`.

## Problem summary

Made specialization for some function calls for its' some columns' ColumnConst. Avoid expanding those columns.

## Performance Improvement

before:
```
mysql [tpch]>Select substring(c_name, 5, 10), substring(c_address, 10, 15), substring(c_comment, 50, 75) from customer;
......
15000000 rows in set (18.30 sec)
```

after:
```
mysql [tpch]>Select substring(c_name, 5, 10), substring(c_address, 10, 15), substring(c_comment, 50, 75) from customer;
......
15000000 rows in set (15.92 sec)
```

## Checklist(Required)

* [ ] Does it affect the original behavior
* [ ] Has unit tests been added
* [ ] Has document been added or modified
* [ ] Does it need to update dependencies
* [x] Is this PR support rollback (If NO, please explain WHY)

## Further comments

There are still some pervious usage haven't update.

### CODE REVIEW

File: be/src/vec/functions/function_bitmap.cpp [hunk 1]
```diff
@@ -452,6 +455,22 @@ struct BitmapNot {
         }
         return Status::OK();
     }
+    static Status vector_const(const TData& lvec, const TData& rvec, TData& res) {
+        size_t size = lvec.size();
+        for (size_t i = 0; i < size; ++i) {
+            res[i] = lvec[i];
+            res[i] -= rvec[0];
```
Comment (HappenLee) [inline be/src/vec/functions/function_bitmap.cpp:462]: directly pass reference， not need [0]

File: be/src/vec/functions/array/function_array_binary.h [hunk 2]
```diff
@@ -51,26 +53,24 @@ class FunctionArrayBinary : public IFunction {
...
                         size_t result, size_t input_rows_count) override {
-        auto left_column =
-                block.get_by_position(arguments[0]).column->convert_to_full_column_if_const();
-        auto right_column =
-                block.get_by_position(arguments[1]).column->convert_to_full_column_if_const();
-        Status ret = Status::RuntimeError(
-                fmt::format("execute failed, unsupported types for function {}({}, {})", get_name(),
-                            block.get_by_position(arguments[0]).type->get_name(),
-                            block.get_by_position(arguments[1]).type->get_name()));
+        auto left_column = unpack_if_const(block.get_by_position(arguments[0]).column);
```
Comment (HappenLee) [inline be/src/vec/functions/array/function_array_binary.h:56]: const auto &

File: be/src/vec/functions/array/function_array_set.h [hunk 3]
```diff
@@ -198,11 +208,23 @@ struct ArraySetImpl {
...
         for (size_t row = 0; row < left_data.offsets_ptr->size(); ++row) {
-            size_t count = 0;
-            size_t left_off = (*left_data.offsets_ptr)[row - 1];
-            size_t left_len = (*left_data.offsets_ptr)[row] - left_off;
-            size_t right_off = (*right_data.offsets_ptr)[row - 1];
-            size_t right_len = (*right_data.offsets_ptr)[row] - right_off;
+            size_t count = 0, left_off, left_len, right_off, right_len;
+            //the `if` can be automatically exchange order with `for` to guarantee performance.
+            if (left_const) {
+                left_off = (*left_data.offsets_ptr)[-1];
```
Comment (HappenLee) [inline be/src/vec/functions/array/function_array_set.h:214]: better move the code out the for loop

Comment (HappenLee) [inline be/src/vec/functions/array/function_array_set.h:214]: better move the code out the for loop

Comment (HappenLee) [inline be/src/vec/functions/array/function_array_set.h:214]: better move the code out the for loop

Comment (HappenLee) [inline be/src/vec/functions/array/function_array_set.h:214]: better move the code out the for loop

File: be/src/vec/columns/column_const.cpp [hunk 4]
```diff
@@ -209,4 +209,11 @@ std::pair<ColumnPtr, size_t> check_column_const_set_readability(const IColumn& c
     }
     return result;
 }
+
+const ColumnPtr& unpack_if_const(const ColumnPtr& ptr) noexcept {
```
Comment (HappenLee) [inline be/src/vec/columns/column_const.cpp:213]: better return std::pair<bool, ColumnPtr&> to reduce unless virtual function call

File: be/src/vec/common/string_ref.cpp [hunk 5]
```diff
@@ -53,6 +53,13 @@ StringRef StringRef::max_string_val() {
     return StringRef((char*)(&StringRef::MAX_CHAR), 1);
 }
 
+bool StringRef::start_with(const char& ch) const {
```
Comment (HappenLee) [inline be/src/vec/common/string_ref.cpp:56]: do not pass origin type with reference

File: be/src/vec/common/string_ref.cpp [hunk 6]
```diff
@@ -53,6 +53,13 @@ StringRef StringRef::max_string_val() {
...
+    return data[0] == ch;
+}
+bool StringRef::end_with(const char& ch) const {
+    return data[size - 1] == ch;
```
Comment (HappenLee) [inline be/src/vec/common/string_ref.cpp:60]: dangerous ！！！！ please check size when you try to access data mem

File: be/src/vec/functions/array/function_array_set.h [hunk 7]
```diff
@@ -150,35 +150,86 @@ struct ArraySetImpl {
     }
 
     static Status execute(ColumnPtr& res_ptr, const ColumnArrayExecutionData& left_data,
-                          const ColumnArrayExecutionData& right_data) {
+                          const ColumnArrayExecutionData& right_data, const bool& left_const,
```
Comment (HappenLee) [inline be/src/vec/functions/array/function_array_set.h:153]: const bool

File: be/src/vec/functions/functions_comparison.h [hunk 8]
```diff
@@ -104,17 +104,18 @@ struct NumComparisonImpl {
...
-    static void NO_INLINE vector_constant(const IColumn& a, const IColumn& b,
-                                          PaddedPODArray<UInt8>& c) {
-        auto b_materialized = b.clone_resized(1)->convert_to_full_column_if_const();
-        for (size_t i = 0, size = a.size(); i < size; ++i)
-            c[i] = Op::apply(a.compare_at(i, 0, *b_materialized, 1), 0);
+    static void vector_constant(const IColumn& a, const IColumn& b, PaddedPODArray<UInt8>& c) {
+        //TODO: double-check this logic. why we use materialization before?
+        for (size_t i = 0, size = a.size(); i < size; ++i) {
+            c[i] = Op::apply(
+                    a.compare_at(i, 0, assert_cast<const ColumnConst&>(b).get_data_column(), 1), 0);
```
Comment (HappenLee) [inline be/src/vec/functions/functions_comparison.h:117]: `assert_cast<const ColumnConst&>(b).get_data_column(), 1` do the work out of for... loop

Comment (HappenLee) [inline be/src/vec/functions/functions_comparison.h:117]: `assert_cast<const ColumnConst&>(b).get_data_column(), 1` do the work out of for... loop

File: be/src/vec/functions/comparison_equal_for_null.cpp [hunk 9]
```diff
@@ -118,14 +115,47 @@ class FunctionEqForNull : public IFunction {
...
             block.get_by_position(result).column = res_nullable_column->get_nested_column_ptr();
         }
         return Status::OK();
     }
+
+private:
+    static void _exec_nullable_equal(unsigned char* result, const unsigned char* left,
+                                     const unsigned char* right, size_t rows, bool left_const,
+                                     bool right_const) {
+        if (left_const && right_const) {
```
Comment (HappenLee) [inline be/src/vec/functions/comparison_equal_for_null.cpp:129]: recheck the function whether can SIMD

File: be/src/vec/functions/function.cpp [hunk 10]
```diff
@@ -166,15 +166,24 @@ Status PreparedFunctionImpl::default_implementation_for_constant_arguments(
...
         const ColumnWithTypeAndName& column = block.get_by_position(args[arg_num]);
-        temporary_block.insert(
-                {assert_cast<const ColumnConst*>(column.column.get())->get_data_column_ptr(),
-                 column.type, column.name});
+        // Columns in const_list --> column_const,    others --> nested_column
+        // that's because some functions supposes some specific columns always constant.
+        // If we unpack it, there will be unnecessary cost of virtual judge.
+        if (args_expect_const.end() !=
+            std::find(args_expect_const.begin(), args_expect_const.end(), arg_num)) {
+            temporary_block.insert({column.column->clone_resized(1), column.type, column.name});
```
Comment (HappenLee) [inline be/src/vec/functions/function.cpp:181]: why clone_resized(1)?

File: be/src/vec/functions/comparison_equal_for_null.cpp (hunk 9 above)
Comment (zclllyybb) [reply be/src/vec/functions/comparison_equal_for_null.cpp:129]: checked for Release version. That's ok for auto-simd

File: be/src/vec/functions/function.cpp (hunk 10 above)
Comment (zclllyybb) [reply be/src/vec/functions/function.cpp]: That's ok to remove it. done.

Comment (zclllyybb) [reply be/src/vec/functions/function.cpp]: done.

**Comment by zclllyybb** [comment]
run buildall

File: be/src/vec/columns/column_const.cpp [hunk 11]
```diff
@@ -209,4 +211,37 @@ std::pair<ColumnPtr, size_t> check_column_const_set_readability(const IColumn& c
...
+                std::cref(static_cast<const ColumnConst&>(*ptr).get_data_column_ptr()), true);
+    }
+    return std::make_pair(std::cref(ptr), false);
+}
+
+void default_preprocess_parameter_columns(ColumnPtr* columns, const bool* col_const,
+                                          const std::initializer_list<size_t>& parameters,
+                                          Block& block, const ColumnNumbers& arg_indexes) noexcept {
+    if (std::all_of(parameters.begin(), parameters.end(),
+                    [&](const size_t& const_index) -> bool { return col_const[const_index]; })) {
```
Comment (HappenLee) [inline be/src/vec/columns/column_const.cpp:227]: do not use reference to pass the origin type like `int/size_t`

File: be/src/vec/functions/array/function_array_set.h [hunk 12]
```diff
@@ -150,35 +150,86 @@ struct ArraySetImpl {
...
+                _execute_internal_rconst<ColumnInt64>(dst, left_data, right_data) ||
+                _execute_internal_rconst<ColumnInt128>(dst, left_data, right_data) ||
+                _execute_internal_rconst<ColumnFloat32>(dst, left_data, right_data) ||
+                _execute_internal_rconst<ColumnFloat64>(dst, left_data, right_data) ||
+                _execute_internal_rconst<ColumnDecimal32>(dst, left_data, right_data) ||
+                _execute_internal_rconst<ColumnDecimal64>(dst, left_data, right_data) ||
+                _execute_internal_rconst<ColumnDecimal128I>(dst, left_data, right_data) ||
+                _execute_internal_rconst<ColumnDecimal128>(dst, left_data, right_data)) {
+                res_column = assemble_column_array(dst);
+                if (res_column) {
```
Comment (HappenLee) [inline be/src/vec/functions/array/function_array_set.h:205]:
```
                res_column = assemble_column_array(dst);
                if (res_column) {
                    res_ptr = std::move(res_column);
                    return Status::OK();
                }
```

the code seems should out of the if / else if / else, be execed in the end

File: be/src/vec/functions/array/function_array_set.h [hunk 13]
```diff
@@ -216,6 +267,66 @@ struct ArraySetImpl {
...
+        if (!check_column<ColumnType>(*left_data.nested_col)) {
+            return false;
+        }
+        constexpr auto execute_left_column_first = Impl::Action::execute_left_column_first;
+        size_t current = 0;
+        Impl impl;
+        for (size_t row = 0; row < left_data.offsets_ptr->size(); ++row) {
+            size_t count = 0;
+            size_t left_off = (*left_data.offsets_ptr)[-1];
+            size_t left_len = (*left_data.offsets_ptr)[0] - left_off;
```
Comment (HappenLee) [inline be/src/vec/functions/array/function_array_set.h:284]: out of the loop to do
```
            size_t left_off = (*left_data.offsets_ptr)[-1];
            size_t left_len = (*left_data.offsets_ptr)[0] - left_off;
```

File: be/src/vec/functions/array/function_array_set.h [hunk 14]
```diff
@@ -216,6 +267,66 @@ struct ArraySetImpl {
...
+            return false;
+        }
+        constexpr auto execute_left_column_first = Impl::Action::execute_left_column_first;
+        size_t current = 0;
+        Impl impl;
+        for (size_t row = 0; row < left_data.offsets_ptr->size(); ++row) {
+            size_t count = 0;
+            size_t left_off = (*left_data.offsets_ptr)[row - 1];
+            size_t left_len = (*left_data.offsets_ptr)[row] - left_off;
+            size_t right_off = (*right_data.offsets_ptr)[-1];
```
Comment (HappenLee) [inline be/src/vec/functions/array/function_array_set.h:315]: out of the loop

**Comment by zclllyybb** [comment]
run buildall

**Pull request closed without merging**